pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -81,10 +89,16 @@ def can_prepare_data(self):
         if self.trainer.datamodule is not None and is_overridden("prepare_data", self.trainer.datamodule):
             should_call_dm_prepare_data = not self.trainer.datamodule._has_prepared_data
 
-        if self.trainer.prepare_data_per_node:
+        if self._prepare_data_per_node():
             return self.trainer.local_rank == 0 and should_call_dm_prepare_data
         return self.trainer.node_rank == 0 and self.trainer.local_rank == 0 and should_call_dm_prepare_data
 
+    def _prepare_data_per_node(self) -> bool:
```

Comment:
this `prepare_data` logic is getting pretty convoluted across these functions

I'd prefer merging `can_prepare_data` into `prepare_data` for now. this will be messy for the timebeing, but will be much simpler in v1.6 when we remove these properties: https://github.com/PyTorchLightning/pytorch-lightning/blob/c6b68883879e38719688865aceac746477f0a9b9/pytorch_lightning/core/datamodule.py#L202-L215

We need to check the `prepare_data_per_node` property on the lightning module too, as the lightning module also inherits from datahooks: https://github.com/PyTorchLightning/pytorch-lightning/blob/c6b68883879e38719688865aceac746477f0a9b9/pytorch_lightning/core/lightning.py#L56-L63

```
def prepare_data(self) -> None:
    # handle datamodule prepare data: check for prepare_data_per_node & datamodule lifecycle properties before calling datamodule.prepare_data
    # handle lightning module prepare data: check for prepare_data_per_node before calling lightning_module.prepare_data
```

```
if self.trainer.data
```

Before calling `
- we need to check if `self.trainer.lightning_module`

Reply:
thanks for the suggestion. I will update that in next iteration.